src/liballoc/collections/vec_deque.rs

```diff
@@ -1019,14 +1019,19 @@ impl<T> VecDeque<T> {
         // the drain is complete and the Drain destructor is run.
         self.head = drain_tail;
 
+        // `deque` and `ring` overlap in what they point to, so we must make sure
+        // that `ring` is "derived-from" `deque`, or else even just creating ring
+        // from `self` already invalidates `deque`.
+        let deque = NonNull::from(&mut *self);
+
         Drain {
-            deque: NonNull::from(&mut *self),
+            deque,
             after_tail: drain_head,
             after_head: head,
             iter: Iter {
                 tail: drain_tail,
                 head: drain_head,
-                ring: unsafe { self.buffer_as_mut_slice() },
+                ring: unsafe { (&mut *deque.as_ptr()).buffer_as_mut_slice() },
```

Comment:
Doesn't `buffer_as_slice()` work here? Why does it need the `_mut_` ?

Reply:
I think there should be a comment here stating that we have to create `ring` from the `deque` pointer to `self` to avoid re-borrowing `self` and invalidating that pointer.

Reply:
I added a comment saying exactly that a few lines above...?

Reply:
> Doesn't buffer_as_slice() work here? Why does it need the _mut_ ? That is, this should work, I think:

Hm, good question. We later actually provide full ownership of the elements based on this iterator, but that's still not mutating anything in there.

And with a shared reference, we don't even need to create a new `&mut self`, so the original order of doing things might even still work!